cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,289 @@
+	// resources.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	PodSets []PodSet `json:"podSets" protobuf:"bytes,1,rep,name=podSets"`
+
+	// ProvisioningClass describes the different modes of provisioning the resources.
+	// Supported values:
+	// * kubernetes.io/checkCapacity - check if current cluster state can fullfil this request, do not reserve the capacity.
+	// * kubernetes.io/atomicScaleUp - provision the resources in an atomic manner
+    // * ... - potential other classes that are specific to the cloud providers
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	ProvisioningClass string `json:"provisioningClass" protobuf:"bytes,2,name=provisioningClass"`
+
+	// AdditionalParameters contains all other parameters custom classes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	AdditionalParameters map[string]string `json:"additionalParameters" protobuf:"bytes,3,name=additionalParameters"`
+}
+
+type PodSet struct {
+	// TemplateRef is a reference to a PodTemplate object that is representing pods 
+	// that will consume this reservation. Users need to make sure that the
+	// fields relevant to scheduler (e.g. node selector tolerations) are
+	// consistent between this template and actual pods consuming the Provisioning Request.
+	TemplateRef *metav1.ObjectReference `json:"templateRef" protobuf:"bytes,1,name=templateRef"`
+	// Count contains the number of pods that will be created with a given
+	// template.
+	// +kubebuilder:validation:Minimum=1
+	Count int32 `json:"count" protobuf:"bytes,2,name=count"`
+}
+
+// ProvisioningRequestStatus represents the status of the resource reservation.
+type ProvisioningRequestStatus struct {
+	// Conditions represent the observations of a Provisioning Request's
+	// current state. Those will contain information whether the capacity
+    // was found/created or if there were any issues. The condition types
+    // may differ between different provisioning classes.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,1,rep,name=conditions"`
+
+	// AdditionalStatus contains all other status values custom provisioning classes may require.
+	//
+	// +optional
+	AdditionalStatus map[string]string `json:"additionalStatus" protobuf:"bytes,2,name=additionalStatus"`
+}
+```
+
+### Provisioning Classes
+
+#### kubernetes.io/checkCapacity class
+
+The `kubernetes.io/checkCapacity` is one-off check to verify that the in the cluster
+there is enough capacity to provision given set of pods.
+
+Note: If two of such objects are created around the same time, CA will consider
+them independently and place no guards for the capacity.
+Also the capacity is not reserved in any manner so it may be scaled-down.
```

Comment:
What are the use cases this intends to solve? I'm concerned that this operations will lead to users expecting outcomes of future requests based on the result of the capacity check.

Reply:
If there is a single actor admitting jobs to the cluster then this can serve as a preliminary check to verify that there is enough capacity in the cluster.